### PR TITLE
Console displaying input/output from other sessions

### LINKED ISSUE #626: Console displaying input/output from kernel from other frontends
When a console interacts with a kernel, it currently only displays input and output originating from that console. It would be useful to have a console option (or default???) that could display *all* input and output from a kernel. Then the console could be a running log of what was evaluated, or let others follow along with what is happening with a kernel if they can connect to the kernel.

@afshin brings up:
> what about if you're connected to the same kernel in an ipython session in your terminal? 
> or what if you're in the middle of an interactive thing like the interactive help?

This came up in conversations with @danielballan and one of his coworkers, in talking about how remote parties can follow along with computation from a notebook or console.

### PR DESCRIPTION
**Pull request by danielballan**
WIP. Closes #626.

Forgive the copious debug log statements -- I'm posting this a bit early so I can ask for a little help.

### CODE REVIEW

File: src/console/widget.ts [hunk 1]
```diff
@@ -134,6 +134,37 @@ class ConsoleWidget extends Widget {
...
+      if (value.parent_header.session === this.session.kernel.clientId) {
+        console.log('message from this session');
+        return;
+      }
+      console.log('message from exernal session');
+      let old_prompt = this.prompt;
+      this.newPrompt();
+      let new_prompt = this.prompt;
+      // Copy any not-yet-executed content into a new prompt.
+      new_prompt.model.source = old_prompt.model.source;
```
Comment (danielballan) [inline src/console/widget.ts]: This works, but...

File: src/console/widget.ts [hunk 2]
```diff
@@ -134,6 +134,37 @@ class ConsoleWidget extends Widget {
...
+      // And fill the old prompt with the content from the remote session.
+      console.log('old_prompt.model before');
+      console.log(old_prompt.model);
+      old_prompt.model.source = value.content.code;
```
Comment (danielballan) [inline src/console/widget.ts]: ...this does not seem to fully work. According to console.log output (below) the source *is* correctly updated, but this is not being reflected in the rendered view. The `source` setter emits a signal that I assume should be triggering a re-draw. Am I missing a step between this and pushing into `this._history` below?

Comment (blink1073) [inline src/console/widget.ts]: I believe the signals for the old prompt are disconnected when you create a new one.
I am burned out from bundling and CSS this week, but can take a deeper look at some other point ;).

Comment (danielballan) [inline src/console/widget.ts]: That would make sense.

Enjoy your weekend. Packaging is exhausting. :- D

**Comment by danielballan** [comment]
@afshin The rebase went fine, I think, but I have one or two questions about `execute_result`. Would you have ~10 minutes sometime tomorrow to briefly chat? My schedule is open. If not we can do this asynchronously, and I'll just expand on my questions here.

**Comment by afshin** [comment]
@danielballan Sure, how about at 10:30AM EST at https://appear.in/jupyter

**Comment by danielballan** [comment]
Sounds good, see you there.

**Comment by danielballan** [comment]
The basic functionality now works.

screen shot 2016-09-17 at 11 58 14 am

One point I am still confused by:

The message from the foreign session has a `msg_type` (`'execute_result'`) but that message's *content* does not have an `output_type` key, as [interface](http://jupyter.org/jupyterlab/interfaces/_notebook_notebook_nbformat_.nbformat.iexecuteresult.html) seems to say it should.

When an output is added to `OutputAreaModel`, it is expected to have an `output_type`. So, here I am providing one manually (line 187 of src/console/content.ts) which I assume is not the intended path. How is `output_type` supposed to enter? This way is clearly not correct; for one thing, it does not correctly represent tracebacks (output_type: error) originating from foreign sessions.

**Comment by blink1073** [comment]
@danielballan, `output_type` comes from the message's `header.msg_type`.

**Comment by danielballan** [comment]
This is cleaned up and ready for another round of review.

In present form, it mirrors 'execute_input' and 'execute_result' but not errors and other outputs. I can see the path to capturing these too, but I would like to run it by someone tomorrow before implementing.

**Comment by blink1073** [comment]
Cool, looks good so far, you're on the right track.

**Comment by danielballan** [comment]
Dark-styled inputs/ouputs are coming from an `ipython console` session attached to the same kernel.

screen shot 2016-09-22 at 10 37 58 am

**Comment by afshin** [comment]
This is awesome, thanks!

**Pull request merged**